### PR TITLE
Synchronized calls to Exporter::Export & Shutdown

### LINKED ISSUE #1158: Synchronized calls to Exporter::Export and Exporter::Shutdown.
As of now, 
1. [SimpleSpanProcessor](https://github.com/open-telemetry/opentelemetry-cpp/blob/48033107c8586c56f21059d603f4ef43f0ce94f3/sdk/include/opentelemetry/sdk/trace/simple_processor.h#L68) ensures that it is not possible to concurrently call Exporter::Shutdown().
2. [SimpleSpanProcessor ](https://github.com/open-telemetry/opentelemetry-cpp/blob/48033107c8586c56f21059d603f4ef43f0ce94f3/sdk/include/opentelemetry/sdk/trace/simple_processor.h#L50) ensures that it is not possible to concurrently call Exporter::Export()
3. But it is still possible to call Exporter::Shutdown() and Exporter::Export() concurrently, and if that happens, is_shutdown/isShutdown variable defined in different exporters need to be protected.

`opentelemetry::common::SpinLockMutex` is lock-free, and one option could be to protect is_shutdown with that.

### PR DESCRIPTION
**Pull request by esigo**
Fixes #1158 (issue)

## Changes

Protects `is_shutdown_` read and writes using `opentelemetry::common::SpinLockMutex`.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by lalitb — APPROVED** [review]

File: exporters/jaeger/src/jaeger_exporter.cc [hunk 1]
```diff
@@ -39,7 +40,7 @@ std::unique_ptr<trace_sdk::Recordable> JaegerExporter::MakeRecordable() noexcept
 sdk_common::ExportResult JaegerExporter::Export(
     const nostd::span<std::unique_ptr<sdk::trace::Recordable>> &spans) noexcept
 {
-  if (is_shutdown_)
+  if (isShutdown())
```
[the remark this answers is not in the document]
Comment (esigo) [reply exporters/jaeger/src/jaeger_exporter.cc:44]: done

File: exporters/jaeger/include/opentelemetry/exporters/jaeger/jaeger_exporter.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 #include <opentelemetry/ext/http/client/http_client.h>
 #include <opentelemetry/sdk/trace/exporter.h>
+#include "opentelemetry/common/spin_lock_mutex.h"
```
Comment (ThomsonTan) [inline exporters/jaeger/include/opentelemetry/exporters/jaeger/jaeger_exporter.h:8]: Use `#include <>` for consistency?

File: exporters/jaeger/src/jaeger_exporter.cc [hunk 3]
```diff
@@ -39,8 +41,9 @@ std::unique_ptr<trace_sdk::Recordable> JaegerExporter::MakeRecordable() noexcept
...
   {
+    OTEL_INTERNAL_LOG_ERROR("[Jaeger Trace Exporter] Export failed, exporter is shutdown");
```
Comment (ThomsonTan) [inline exporters/jaeger/src/jaeger_exporter.cc:46]: Log the size of `span` to imply how many spans are lost?

File: exporters/jaeger/src/jaeger_exporter.cc [hunk 4]
```diff
@@ -91,10 +94,17 @@ void JaegerExporter::InitializeEndpoint()
 
 bool JaegerExporter::Shutdown(std::chrono::microseconds timeout) noexcept
 {
+  const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
   is_shutdown_ = true;
   return true;
 }
 
+const bool JaegerExporter::isShutdown() const noexcept
```
Comment (ThomsonTan) [inline exporters/jaeger/src/jaeger_exporter.cc:102]: No need to return const as the returned value is always a copy?

Comment (esigo) [reply exporters/jaeger/src/jaeger_exporter.cc]: yeah thanks, no need for `const`.

File: exporters/jaeger/src/jaeger_exporter.cc (hunk 3 above)
Comment (esigo) [reply exporters/jaeger/src/jaeger_exporter.cc]: added where possible

File: exporters/jaeger/include/opentelemetry/exporters/jaeger/jaeger_exporter.h (hunk 2 above)
Comment (esigo) [reply exporters/jaeger/include/opentelemetry/exporters/jaeger/jaeger_exporter.h]: done

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**